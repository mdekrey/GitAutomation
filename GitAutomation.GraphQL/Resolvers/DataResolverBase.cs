using GraphQL;
using GraphQL.Types;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GitAutomation.GraphQL.Resolvers
{
    abstract class DataResolverBase<TSource, TIn, TOut> : IDataResolver<TSource>
        where TIn : class, new()
    {
        public Task<object> Resolve(ResolveFieldContext<TSource> context)
        {
            var argInstance = GetArguments(context);
            return Resolve(argInstance).ContinueWith(task => (object)task.Result);
        }

        protected virtual TIn GetArguments(ResolveFieldContext<TSource> context)
        {
            return context.Arguments.ToObject<TIn>();
        }

        protected abstract Task<TOut> Resolve(TIn args);
    }

}
