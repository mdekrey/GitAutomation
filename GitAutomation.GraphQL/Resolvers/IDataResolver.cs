using GraphQL.Types;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GitAutomation.GraphQL.Resolvers
{
    interface IDataResolver<TSource>
    {
        Task<object> Resolve(ResolveFieldContext<TSource> context);
    }

}
