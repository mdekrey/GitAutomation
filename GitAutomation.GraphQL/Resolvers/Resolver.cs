using GraphQL.Resolvers;
using System;
using System.Collections.Generic;
using System.Text;
using GraphQL.Types;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using System.Linq.Expressions;
using System.Linq;
using System.Reflection;

namespace GitAutomation.GraphQL.Resolvers
{
    public class Resolver : IFieldResolver
    {
        private Func<ResolveFieldContext, Task<object>> func;

        private Resolver(Func<ResolveFieldContext, Task<object>> func)
        {
            this.func = func;
        }
        
        public static IFieldResolver Resolve(object target, string methodName)
        {
            if (target == null)
            {
                throw new ArgumentNullException(nameof(target));
            }
            var method = target.GetType().GetMethod(methodName, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
            if (method == null)
            {
                throw new ArgumentException("Method not found on target.", nameof(methodName));
            }

            var paramTypes = (from param in method.GetParameters()
                              select param.ParameterType).ToArray();

            if (!method.ReturnType.IsConstructedGenericType || method.ReturnType.GetGenericTypeDefinition() != typeof(Task<>))
            {
                throw new ArgumentException("Return type must be of type Task<>", nameof(methodName));
            }

            var contextParameter = Expression.Parameter(typeof(ResolveFieldContext), "context");
            var arguments = (from param in method.GetParameters()
                             select ParameterToExpression(param, contextParameter)).ToArray();
            var result = Expression.Lambda<Func<ResolveFieldContext, Task<object>>>(
                Expression.Call(
                    Expression.Call(
                        instance: Expression.Constant(target, target.GetType()),
                        method: method,
                        arguments: arguments
                    ),
                    nameof(Task.ContinueWith),
                    new Type[] { typeof(object) },
                    Expression.Constant(Delegate.CreateDelegate(typeof(Func<,>).MakeGenericType(method.ReturnType, typeof(object)), FinalCastMethod.MakeGenericMethod(method.ReturnType.GetGenericArguments())))
                ),
                contextParameter
            );

            return new Resolver(result.Compile());
        }

        private static Expression ParameterToExpression(ParameterInfo param, ParameterExpression contextParameter)
        {
            if (param.IsOut || param.IsRetval)
            {
                throw new ArgumentException("Invalid argument; parameter cannot be out or ref.", param.Name);
            }
            var attr = param.GetCustomAttribute<ResolverAttribute>();
            switch (attr)
            {
                case SourceAttribute a:
                    return Expression.Convert(Expression.Property(contextParameter, nameof(ResolveFieldContext.Source)), param.ParameterType);

                case FromServicesAttribute a:
                    var serviceProviderExpression = Expression.Convert(
                            Expression.Property(
                                contextParameter,
                                nameof(ResolveFieldContext.UserContext)
                            ),
                            typeof(IServiceProvider)
                        );
                    return Expression.Call(
                        (param.IsOptional
                            ? GetOptionalServiceMethod
                            : GetRequiredServiceMethod)
                            .MakeGenericMethod(param.ParameterType),
                        serviceProviderExpression
                    );

                case FromArgumentAttribute a:
                    return Expression.Call(contextParameter, nameof(ResolveFieldContext.GetArgument), new[] { param.ParameterType }, Expression.Constant(param.Name));

                default:
                    throw new ArgumentException("Invalid argument; parameter must have a ResolverAttribute.", param.Name);
            }

        }

        private static readonly MethodInfo GetOptionalServiceMethod = typeof(Resolver).GetMethod(nameof(GetOptionalService), BindingFlags.Static | BindingFlags.NonPublic);
        private static T GetOptionalService<T>(IServiceProvider serviceProvider)
        {
            return serviceProvider.GetService<T>();
        }

        private static readonly MethodInfo GetRequiredServiceMethod = typeof(Resolver).GetMethod(nameof(GetRequiredService), BindingFlags.Static | BindingFlags.NonPublic);
        private static T GetRequiredService<T>(IServiceProvider serviceProvider)
        {
            return serviceProvider.GetRequiredService<T>();
        }

        private static readonly MethodInfo FinalCastMethod = typeof(Resolver).GetMethod(nameof(FinalCast), BindingFlags.Static | BindingFlags.NonPublic);
        private static object FinalCast<T>(Task<T> input)
        {
            return input.Result;
        }

        public object Resolve(ResolveFieldContext context)
        {
            return func(context);
        }
    }
}
