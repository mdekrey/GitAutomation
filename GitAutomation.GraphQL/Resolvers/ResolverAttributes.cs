using System;
using System.Collections.Generic;
using System.Text;

namespace GitAutomation.GraphQL.Resolvers
{
    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = true)]
    public abstract class ResolverAttribute : Attribute
    {
    }

    public class SourceAttribute : ResolverAttribute
    {
    }

    public class FromArgumentAttribute : ResolverAttribute
    {
        public string Name { get; set; }
        public string Description { get; set; }
    }

    public class FromServicesAttribute : ResolverAttribute
    {
    }
}
