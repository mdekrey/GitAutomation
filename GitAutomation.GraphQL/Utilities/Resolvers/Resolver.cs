using GraphQL.Resolvers;
using System;
using System.Collections.Generic;
using System.Text;
using GraphQL.Types;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using System.Linq.Expressions;
using System.Linq;
using System.Reflection;
using GraphQL;
using Microsoft.Extensions.Logging;

namespace GitAutomation.GraphQL.Utilities.Resolvers
{
    public class Resolver : IFieldResolver
    {
        private Func<ResolveFieldContext, object> func;

        private Resolver(Func<ResolveFieldContext, object> func)
        {
            this.func = func;
        }

        public static IFieldResolver Resolve(object target, string methodName)
        {
            if (target == null)
            {
                throw new ArgumentNullException(nameof(target));
            }
            var method = target.GetType().GetMethod(methodName, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
            if (method == null)
            {
                throw new ArgumentException("Method not found on target.", nameof(methodName));
            }

            var paramTypes = (from param in method.GetParameters()
                              select param.ParameterType).ToArray();

            if (!method.ReturnType.IsConstructedGenericType || method.ReturnType.GetGenericTypeDefinition() != typeof(Task<>))
            {
                throw new ArgumentException("Return type must be of type Task<>", nameof(methodName));
            }

            var contextParameter = Expression.Parameter(typeof(ResolveFieldContext), "context");
            var arguments = (from param in method.GetParameters()
                             select ParameterToExpression(param, contextParameter)).ToArray();
            var result = Expression.Lambda<Func<ResolveFieldContext, object>>(
                Expression.Convert(
                    Expression.Call(
                        instance: Expression.Constant(target, target.GetType()),
                        method: method,
                        arguments: arguments
                    ),
                    typeof(object)
                ),
                contextParameter
            );

            return new Resolver(result.Compile());
        }

        private static Expression ParameterToExpression(ParameterInfo param, ParameterExpression contextParameter)
        {
            if (param.IsOut || param.IsRetval)
            {
                throw new ArgumentException("Invalid argument; parameter cannot be out or ref.", param.Name);
            }
            var attr = param.GetCustomAttribute<ResolverAttribute>();
            switch (attr)
            {
                case SourceAttribute a:
                    return Expression.Convert(Expression.Property(contextParameter, nameof(ResolveFieldContext.Source)), param.ParameterType);

                case FromServicesAttribute a:
                    var serviceProviderExpression = Expression.Convert(
                            Expression.Property(
                                contextParameter,
                                nameof(ResolveFieldContext.UserContext)
                            ),
                            typeof(IServiceProvider)
                        );
                    return Expression.Call(
                        (param.IsOptional
                            ? GetOptionalServiceMethod
                            : GetRequiredServiceMethod)
                            .MakeGenericMethod(param.ParameterType),
                        serviceProviderExpression
                    );

                case FromArgumentAttribute a:
                    return Expression.Call(
                        contextParameter,
                        GetArgumentGeneric.MakeGenericMethod(param.ParameterType),
                        Expression.Constant(param.Name),
                        Expression.Constant(
                            param.ParameterType.IsValueType
                                ? Activator.CreateInstance(param.ParameterType)
                                : null,
                            param.ParameterType
                        )
                    );

                default:
                    if (param.ParameterType == contextParameter.Type)
                    {
                        return contextParameter;
                    }
                    throw new ArgumentException("Invalid argument; parameter must have a ResolverAttribute.", param.Name);
            }

        }

        private static readonly MethodInfo GetArgumentGeneric =
            typeof(ResolveFieldContext).GetMethods()
                .Where(m => m.Name == nameof(ResolveFieldContext.GetArgument) && m.IsGenericMethodDefinition)
                .SingleOrDefault();

        private static readonly MethodInfo GetOptionalServiceMethod = typeof(Resolver).GetMethod(nameof(GetOptionalService), BindingFlags.Static | BindingFlags.NonPublic);
        private static T GetOptionalService<T>(IServiceProvider serviceProvider)
        {
            return serviceProvider.GetService<T>();
        }

        private static readonly MethodInfo GetRequiredServiceMethod = typeof(Resolver).GetMethod(nameof(GetRequiredService), BindingFlags.Static | BindingFlags.NonPublic);
        private static T GetRequiredService<T>(IServiceProvider serviceProvider)
        {
            return serviceProvider.GetRequiredService<T>();
        }

        private static readonly MethodInfo FinalCastMethod = typeof(Resolver).GetMethod(nameof(FinalCast), BindingFlags.Static | BindingFlags.NonPublic);
        private static object FinalCast<T>(Task<T> input)
        {
            return input.Result;
        }

        public object Resolve(ResolveFieldContext context)
        {
            return ResolveAsync(context);
        }

        public async Task<object> ResolveAsync(ResolveFieldContext context)
        {
            try
            {
                var logger = (context.UserContext as IServiceProvider).GetRequiredService<ILoggerFactory>().CreateLogger<Resolver>();
                logger.LogInformation("Resolving field {0}", context.FieldName);
                var result = func(context);
                if (result is Task t)
                {
                    await t.ConfigureAwait(false);
                    result = t.GetProperyValue(nameof(Task<object>.Result));
                }
                logger.LogInformation("Finished resolving field {0}", context.FieldName);
                return result;
            }
            catch (Exception ex)
            {
                var err = new ExecutionError(ex.Message);
                err.AddLocation(context.FieldAst, context.Document);
                context.Errors.Add(err);
                return null;
            }
        }
    }
}
