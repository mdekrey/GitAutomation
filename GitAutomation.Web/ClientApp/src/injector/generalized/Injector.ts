import { Resolvers } from "./Resolvers";

export class Injector<TServices extends {}, TScope extends string> {
  private readonly serviceResolvers: Resolvers<TServices>;
  private readonly scopeByService: Record<keyof TServices, TScope | null>;
  private readonly scopedResolutions: {
    [scopeLevel in TScope]?: Partial<TServices>
  };
  private readonly scopeHierarchy: TScope[];

  constructor(
    serviceResolvers: Resolvers<TServices>,
    scopeByService: Record<keyof TServices, TScope | null>,
    scopeHierarchy: TScope[],
    scopedResolutions: { [scopeLevel in TScope]?: Partial<TServices> } = {}
  ) {
    this.serviceResolvers = serviceResolvers;
    this.scopeByService = scopeByService;
    this.scopeHierarchy = scopeHierarchy;
    this.scopedResolutions = scopedResolutions;
  }

  beginScope(scope: TScope) {
    this.scopedResolutions[scope] = {};
  }
  endScope(scope: TScope) {
    delete this.scopedResolutions[scope];
  }

  createChildInjector() {
    const parent = this.scopedResolutions;
    const scopedResolutions: typeof parent = {};
    Object.setPrototypeOf(scopedResolutions, parent);
    return new Injector(
      this.serviceResolvers,
      this.scopeByService,
      this.scopeHierarchy,
      scopedResolutions
    );
  }

  resolve<TService extends keyof TServices>(
    service: TService,
    inputMinScope?: TScope
  ): TServices[TService] {
    const minScope =
      inputMinScope === undefined
        ? this.scopeHierarchy[this.scopeHierarchy.length - 1]
        : inputMinScope;
    const minScopeLevel = this.scopeHierarchy.indexOf(minScope);
    for (const scope of this.scopeHierarchy) {
      if (this.scopeHierarchy[minScopeLevel + 1] === scope) {
        // extended out of scope
        break;
      }
      const scopedResolutions = this.scopedResolutions[scope];
      if (scopedResolutions) {
        if (service in scopedResolutions) {
          return scopedResolutions[service]!;
        }
      }
    }

    const targetScope = this.scopeByService[service];
    const targetScopeLevel = this.getScopeLevel(targetScope);
    if (targetScopeLevel !== null && targetScopeLevel > minScopeLevel) {
      throw new Error(
        `Target scope '${targetScope}' is more specific than allowed '${minScope}' for service '${service}'.`
      );
    }
    if (targetScope !== null && !this.scopedResolutions[targetScope!]) {
      throw new Error(
        `Target scope '${targetScope}' has not been started for service '${service}'.`
      );
    }
    const nextScope =
      targetScopeLevel !== null && targetScopeLevel < minScopeLevel
        ? targetScope!
        : minScope;
    try {
      const result = this.serviceResolvers[service](newService =>
        this.resolve(newService, nextScope)
      );
      if (targetScope !== null) {
        const scopedResolutions: Partial<TServices> = this.scopedResolutions[
          targetScope!
        ]!;
        scopedResolutions[service] = result;
      }
      return result;
    } catch (err) {
      throw new Error(
        `When resolving Service '${service}' in Scope '${targetScope}'...\n  ${err}`
      );
    }
  }

  private getScopeLevel(scope: TScope | null): number | null {
    if (scope === null) {
      return null;
    }
    const scopeLevel = this.scopeHierarchy.indexOf(scope);
    if (scopeLevel === -1) {
      throw new Error(
        `Scope '${scope}' not in hierarchy '${this.scopeHierarchy}'`
      );
    }
    return scopeLevel;
  }
}
