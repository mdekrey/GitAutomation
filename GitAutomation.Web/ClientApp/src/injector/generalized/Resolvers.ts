import { Resolver } from "./Resolver"

export type Resolvers<TServices extends {}> = {
  [service in keyof TServices]: Resolver<TServices, TServices[service]>
}
