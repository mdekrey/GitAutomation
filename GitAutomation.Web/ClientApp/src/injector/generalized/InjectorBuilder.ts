import { Injector } from "./Injector";
import { Resolvers } from "./Resolvers";
import { Resolver } from "./Resolver";

export class InjectorBuilder<TServices extends {}, TScope extends string> {
  private readonly scopeHierarchy: TScope[];
  private readonly serviceResolvers: Partial<Resolvers<TServices>>;
  private readonly scopeByService: Partial<
    Record<keyof TServices, TScope | null>
  >;

  constructor(
    scopeHierarchy: TScope[],
    serviceResolvers: Partial<Resolvers<TServices>> = {},
    scopeByService: Partial<Record<keyof TServices, TScope | null>> = {}
  ) {
    this.scopeHierarchy = scopeHierarchy;
    this.serviceResolvers = serviceResolvers;
    this.scopeByService = scopeByService;
  }

  set<TService extends keyof TServices>(
    service: TService,
    scope: TScope | null,
    resolver: Resolver<TServices, TServices[TService]>
  ) {
    this.serviceResolvers[service] = resolver;
    this.scopeByService[service] = scope;
    return this;
  }

  getResolver<TService extends keyof TServices>(service: TService) {
    return this.serviceResolvers[service];
  }

  getScope<TService extends keyof TServices>(service: TService) {
    return this.scopeByService[service];
  }

  copy() {
    return new InjectorBuilder(
      this.scopeHierarchy,
      { ...this.serviceResolvers },
      { ...this.scopeByService }
    );
  }

  build(): Injector<TServices, TScope> {
    return new Injector(
      this.serviceResolvers as Resolvers<TServices>,
      this.scopeByService as Record<keyof TServices, TScope | null>,
      this.scopeHierarchy
    );
  }
}
