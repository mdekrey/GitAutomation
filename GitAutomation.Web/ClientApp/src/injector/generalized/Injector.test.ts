import { Injector } from "./Injector";
import { Resolvers } from "./Resolvers";
import { InjectorBuilder } from "./InjectorBuilder";

const wheel: unique symbol = Symbol();
class Wheel {
  public readonly wheel = wheel;
}
class Vehicle {
  constructor(public readonly track: Track, public readonly wheels: Wheel[]) {}
}
const track: unique symbol = Symbol();
class Track {
  public readonly track = track;
}

enum Scope {
  Singleton = "Singleton",
  Race = "Race",
}
const scopeHierarchy = [Scope.Singleton, Scope.Race];

interface Services {
  wheel: Wheel;
  motorcycleFactory: () => Vehicle;
  carFactory: () => Vehicle;
  playerVehicle: Vehicle;
  opponentVehicle: Vehicle;
  track: Track;
  contestants: Vehicle[];
}

const builder = new InjectorBuilder<Services, Scope>(scopeHierarchy)
  .set("wheel", null, () => new Wheel())
  .set("motorcycleFactory", null, resolver => () =>
    new Vehicle(resolver("track"), [resolver("wheel"), resolver("wheel")])
  )
  .set("carFactory", null, resolver => () =>
    new Vehicle(resolver("track"), [
      resolver("wheel"),
      resolver("wheel"),
      resolver("wheel"),
      resolver("wheel"),
    ])
  )
  .set("playerVehicle", Scope.Singleton, resolver =>
    resolver("motorcycleFactory")()
  )
  .set("opponentVehicle", Scope.Race, resolver => resolver("carFactory")())
  .set("track", Scope.Singleton, resolver => new Track())
  .set("contestants", null, resolver => [
    resolver("playerVehicle"),
    resolver("opponentVehicle"),
  ]);

it("can create an injector", async () => {
  const target = builder.build();
});

it("can resolve transients without starting a scope", async () => {
  const target = builder.build();
  const actual = target.resolve("wheel");
  expect(actual).toBeInstanceOf(Wheel);
});

it("resolves transients each time", async () => {
  const target = builder.build();
  const actual = target.resolve("wheel");
  const actual2 = target.resolve("wheel");
  expect(actual).not.toBe(actual2);
  expect(actual).toBeInstanceOf(Wheel);
  expect(actual2).toBeInstanceOf(Wheel);
});

it("cannot resolve a scoped variable before starting the scope", async () => {
  const target = builder.build();
  expect(() => target.resolve("playerVehicle")).toThrowErrorMatchingSnapshot();
});

it("can resolve a scoped variable after starting the scope", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const actual = target.resolve("playerVehicle");
  expect(actual).toBeInstanceOf(Vehicle);
});

it("can resolve a scoped variable as the same repeatedly", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const actual = target.resolve("playerVehicle");
  const actual2 = target.resolve("playerVehicle");
  expect(actual).toBe(actual2);
});

it("can have multiple scopes", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
});

it("resolves similar types differently", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
  const actual = target.resolve("playerVehicle");
  const actual2 = target.resolve("opponentVehicle");
  expect(actual).not.toBe(actual2);
});

it("resolves a different variable after ending scope and restarting", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
  const actual = target.resolve("opponentVehicle");
  target.endScope(Scope.Race);
  target.beginScope(Scope.Race);
  const actual2 = target.resolve("opponentVehicle");
  expect(actual).not.toBe(actual2);
});

it("cannot resolve a service after ending scope", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
  target.resolve("opponentVehicle");
  target.endScope(Scope.Race);
  expect(() =>
    target.resolve("opponentVehicle")
  ).toThrowErrorMatchingSnapshot();
});

it("cannot resolve a service in a lower scope", async () => {
  const target = builder
    .copy()
    .set("track", Scope.Race, builder.getResolver("track"))
    .build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
  expect(() => target.resolve("playerVehicle")).toThrowErrorMatchingSnapshot();
});

it("transients can resolve services in any scope", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  target.beginScope(Scope.Race);
  const contestants = target.resolve("contestants");
  const opponent = target.resolve("opponentVehicle");
  const player = target.resolve("playerVehicle");
  expect(contestants).toContain(opponent);
  expect(contestants).toContain(player);
});

it("shares objects with parent injectors scopes", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const child1 = target.createChildInjector();
  const player1 = child1.resolve("playerVehicle");
  const child2 = target.createChildInjector();
  const player2 = child2.resolve("playerVehicle");
  expect(player1).toBe(player2);
});

it("can have separate scopes from child injectors", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const child1 = target.createChildInjector();
  child1.beginScope(Scope.Race);
  expect(() =>
    target.resolve("opponentVehicle")
  ).toThrowErrorMatchingSnapshot();
});

it("preserves own scope when child injectors are used", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const child1 = target.createChildInjector();
  child1.beginScope(Scope.Race);
  const opponent1 = child1.resolve("opponentVehicle");
  const child2 = target.createChildInjector();
  child2.beginScope(Scope.Race);
  const opponent2 = child1.resolve("opponentVehicle");
  expect(opponent1).toBe(opponent2);
});

it("can have separate scopes when child injectors are used", async () => {
  const target = builder.build();
  target.beginScope(Scope.Singleton);
  const child1 = target.createChildInjector();
  child1.beginScope(Scope.Race);
  const opponent1 = child1.resolve("opponentVehicle");
  const child2 = target.createChildInjector();
  child2.beginScope(Scope.Race);
  const opponent2 = child2.resolve("opponentVehicle");
  expect(opponent1).not.toBe(opponent2);
  expect(opponent1.track).toBe(opponent2.track);
});
