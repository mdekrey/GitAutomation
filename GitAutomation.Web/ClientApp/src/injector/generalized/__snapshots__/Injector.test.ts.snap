// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can have separate scopes from child injectors 1`] = `"Target scope 'Race' has not been started for service 'opponentVehicle'."`;

exports[`cannot resolve a scoped variable before starting the scope 1`] = `"Target scope 'Singleton' has not been started for service 'playerVehicle'."`;

exports[`cannot resolve a service after ending scope 1`] = `"Target scope 'Race' has not been started for service 'opponentVehicle'."`;

exports[`cannot resolve a service in a lower scope 1`] = `
"When resolving Service 'playerVehicle' in Scope 'Singleton'...
  Error: Target scope 'Race' is more specific than allowed 'Singleton' for service 'track'."
`;
